### DISCUSSION TITLE
IP Pool direct on postfix

### DISCUSSION THREAD

**Discussion by leoborlot** [docker-mailserver/docker-mailserver#4102] [Q&A]
Hi guys, any idea how config an IP Pool direct at postfix?
My server have a 64ip on vEth.

**Comment by polarathene** [comment]
Could you elaborate a bit more? What is your actual goal / benefit you're wanting to accomplish from such? Is there an actual problem you're trying to solve?

From what I've seen, you can use Traefik (especially with kubernetes) to load balance those IPs to internal services, or have a single IP inbound that load balances to different IP of  your internal network / containers.

Postfix and Dovecot though in DMS are not configured for clustering, there was some efforts by users in the past, but none managed to contribute any of that support upstream into DMS. I am not able to advise on this, other than we develop DMS without cluster support in mind, you may run into issues there if that is what you're seeking.

If it was multiple IP to the same instance, I'm not sure what that has to do with DMS, you might have some issues with delivery due to rDNS, I'm not sure.

**Comment by leoborlot** [comment]
My server has a block of 128 IPs. I use these IPs for an exclusive SMTP application, where I configure each email to be sent from a different IP, forming an IP pool. This ensures the rotation of sends and prevents all emails from being penalized if one of the IPs is compromised.

I am not talking about incoming IPs, but outgoing ones.
I saw that I can use an external relay to take advantage of the IP pool, but I would like to know if it is possible to configure this directly in Postfix.

**Reply by polarathene** [reply]
> This ensures the rotation of sends and prevents all emails from being penalized if one of the IPs is compromised.

Why would one of the IP become compromised? Wouldn't the service itself be compromised? Then the rotation would just poison the rest of your IP pool?

Is this IPv6? I don't know how penalizing an IP is approached, but would they not recognize the IP block reputation becoming bad and just associate the whole block? I can't recall IPv6 subnet ranges that well, but considering the much larger pool of addresses I wouldn't be surprised if a penalty was applied to a wider prefix that is typically assigned to a server?

Is this actually a best practice you have come across somewhere? Are you facing an actual issue with your outbound IP addresses becoming penalized/compromised? Might be better to identify why that is?

---

> I am not talking about incoming IPs, but outgoing ones.

You can bind the outbound `smtp` Postfix service to a specific interface: https://docker-mailserver.github.io/docker-mailserver/latest/examples/use-cases/bind-smtp-network-interface/#alternative-unverified

To bind multiple, you could probably adopt advice from here: https://docker-mailserver.github.io/docker-mailserver/latest/examples/tutorials/mailserver-behind-proxy/#dms-postfix-dovecot

You would then need to add some additional Postfix config with some logic for which one of those `smtp` service to deliver mail outbound with (_[`transport_maps` for selecting by recipient address](https://www.postfix.org/postconf.5.html#transport_maps), while for selection by sender address should be via [`sender_dependent_default_transport_maps`](https://www.postfix.org/postconf.5.html#sender_dependent_default_transport_maps)_). Configure that with our `postfix-main.cf` and volume mount the [`transport` table](https://www.postfix.org/transport.5.html) you configure for the setting you added in `postfix-main.cf`.

DMS does not officially support such, you'll have to figure that out yourself beyond the above advice 👍 (_I have no clue how you'd go about rotating outbound IP instead of via transport maps_)

---

You could possibly use the advice in the interface binding docs guide I linked above, and have `postfix-main.cf` / `postfix-master.cf` file updated by your own external script, cycling the IP address to bind with at some interval.

DMS `check-for-changes.sh` feature should detect the updated config then and apply it. Don't cycle too quickly, every 30-60s is probably the lowest interval you should attempt.
